Replace old mod overlay with new design

bdach opened the pull request:
#16917 continued.

Well, here it is. It's only taken what, almost 3 months?

https://user-images.githubusercontent.com/20418176/167026217-9b669fc9-dd0c-49f7-9813-94d15cc5ace7.mp4

I'm PRing this because it "mostly works", but I'm not sure how well it works to others and meets expectations. So this is just mostly being PRed because I have a diff to be PRed and it looks to be working, but there are contention points and I am expecting this to take days (if not weeks and several detours) before merge.

Commits are split such that pulling parts of this into separate PRs should not be too difficult, if deemed required - but all things considered, the diffstat is still well under 500 lines.

## Contention points

### The introduction of `IOverlayManager`

As design has it, the mod overlay becomes a fullscreen overlay rather than being the "odd one out" thus far - the old design was firmly contained within song select which was - I believe - the only case of that. For multiple reasons [outlined in this comment](https://github.com/ppy/osu/issues/16917#issuecomment-1117928322), I need to escape the song select screen and register at game level from within a screen, which is a new use case that `IOverlayManager` facilitates.

I've managed to salvage testability by implementing `IOverlayManager` in `ScreenTestScene` too, so it shouldn't be all bad from that standpoint. But the API of `IOverlayManager` could well be called into question.

Before you ask, `OverlayActivationMode` was originally not supposed to be in `IOverlayManager`, but `OsuFocusedOverlayContainer` uses that, and I really didn't want to multiply interfaces if I could avoid it.

### Some satellite test fixes

Other tests were depending on some bits and bobs of the old overlay, and failures only surfaced upon replacement. Most of it is straightforward class substitution, aside from:

* 17c19063f7945ad3d40287f471875c0a109835f4: Test scene kept appending copies of `OsuModNoFail` to `SelectedMods` shared by entirety of test scene which would fail invariants held by the new mod select that the old one didn't care about. Fixed by fixing the test.
* d296f78dfe097fdf70c759a416538ef7a1fa3fc5: Needed some additional jiggling to test that the buttons are present correctly and to make sure that the buttons aren't checked too early.
* a2ab79620a2a7dd63260a23951b9b2d572046d56: Was relying on back button to be present (see below).

### The song select footer is gone

The old overlay would show the song select footer, the new one doesn't. Partly because that's just how it's come out after moving the overlay out to `OsuGame` level, partly because I was intensely confused why there were at least five disparate methods to exit the screen as it was presented [on the figma designs](https://www.figma.com/file/NkdREXr6wFkA99oLw8PnBv/Client%2FMod-Selection?node-id=0%3A1), namely: click dimmed area, click "X" button on top right, click "back" footer button, click  "mods" footer button, click "confirm" footer button. I've preserved the first three for the time being.

### Mod overlay needs to handle the F1 hotkey to close itself

See 34cf4c6a3895557d6dc5f1d07b702857b512ccca; consequence of registering the overlay at game level as it takes keyboard focus away from the stack. Another one of these things you wouldn't notice before the thing being put inside the actual game.

### Mod overlays in online screens use the plum theme to match the rest of online

Wasn't really required and I had to go out of my way a non-negligible amount to accommodate this, but felt nice to do. Can split out on request, but it really doesn't take too many lines of the total diff.

### User mod overlay in multiplayer doesn't hide empty columns

This is more tricky to achieve than you would think because of the pop in transition (wherein columns move in from both sides of the screen). I've tried once but failed, can try again if it's really wanted.

## Missing pieces / satellite changes

### Resolved

- [x] https://github.com/ppy/osu/pull/18119
- [x] https://github.com/ppy/osu/pull/18122
- [x] https://github.com/ppy/osu/pull/18126
- [x] https://github.com/ppy/osu/pull/18130
- [x] https://github.com/ppy/osu/pull/18131
- [x] https://github.com/ppy/osu/pull/18132
- [x] https://github.com/ppy/osu/pull/18133
- [x] Select binding should dismiss the overlay: Applied in this pull

### Left as follow-up

- Consider hiding toolbar upon showing mod select
  - Can be done, but I'd like to build on top of `IOverlayManager` for this
- Consider scrolling to selections made using keyboard
  - The old design didn't do this
- Fix sidebar overlays not working properly with the mod select
  - This is [already the case for other overlays](https://github.com/ppy/osu/pull/18111#issuecomment-1119892511) and was somewhat improved by https://github.com/ppy/osu/pull/18125
- Mod panel samples play when ctrl-enter binding is pressed
  - Already have changes ready for this

bdach committed:
Replace old mod select overlay with new design

bdach committed:
Fix new mod select overlay dimming itself

bdach committed:
Use plum colour scheme in multiplayer to match rest of multiplayer screens

bdach committed:
Use new free mod select design in room creation flow

bdach committed:
Use new user mod select design in multiplayer lobby

bdach committed:
Extract interface for overlay management

bdach committed:
Implement `IOverlayManager` in `ScreenTestScene`

bdach committed:
Fix mod selects not hiding when their owner screens exit

bdach committed:
Add failing test coverage for toggling mod overlay with hotkey

bdach committed:
Fix mod overlay not closing on toggle hotkey

bdach committed:
Fix user dim test scene accumulating no fail instances in `SelectedMods`

The old design was apparently somehow tolerant of multiple instances of
the same mod in `SelectedMods`, but the new one is not.

Fix the test, because appending to a list of mods that was getting
modified by previous tests is generally not how tests should be
written.

bdach committed:
Replace reference to old mod select in multiplayer test

bdach committed:
Replace references to old mod select classes in match song select test

Also note the change from `songSelect.ChildrenOfType<>()...` to
`this.ChildrenOfType<>()...` - because the new design is registered at
game-level, the mod select is not a child of the song select screen
anymore.

bdach committed:
Replace references to old mod select classes in multiplayer match screen test

bdach committed:
Update song select exit via click test to match new expectations

Joehuu commented:
Not sure if this was an oversight, but the "deselect all" and "select all (mod sections)"  (for freemods) buttons and their respective shortcuts are completely missing.

With the foot obstructed, you can't see the "mods" button that displays all your selected mods in one spot. If I were to accidentally pick a mod in the far-right like "wiggle" and scroll back to the left, I wouldn't know until closing the overlay.

bdach commented:
> "deselect all" and "select all (mod sections)" (for freemods) buttons and their respective shortcuts are completely missing

I did forget about that (or rather, wasn't entirely aware that this was a thing), yes. I'll look into adding that back in a separate pull.

> you can't see the "mods" button that displays all your selected mods in one spot

Fair, but I'll note that the figma designs give zero consideration for this, too. I'd have to wing something.

peppy commented:
A few notes:

- I'd like to see a "back" button added (in a similar style to the one in the first run overlay) to the left of the customisation button, to allow users to get back to song select by clicking roughly where they clicked to enter the dialog. One would hope that in the future this is one and the same button (ie. the mod overlay toggle button is also visible on song select and stays in place) but this should do for now
- Need to add back select/deselect all mods in the freemod select mod. I think this is pretty important as the most common scenario is enabling all mods for freemod, which currently requires clicking on each column's select-all button (so five times or so).

bdach commented:
I'll also look into other things mentioned offhand, such as:

* Hiding the toolbar when the new mod overlay is shown.
* If a mod off-screen is selected using keyboard, the view should scroll to it.

peppy commented:
Also propose that the select binding should dismiss the overlay:

```diff
diff --git a/osu.Game/Overlays/Mods/ModSelectScreen.cs b/osu.Game/Overlays/Mods/ModSelectScreen.cs
index 7c0ed941c6..10f8b352e6 100644
--- a/osu.Game/Overlays/Mods/ModSelectScreen.cs
+++ b/osu.Game/Overlays/Mods/ModSelectScreen.cs
@@ -340,7 +340,7 @@ public override bool OnPressed(KeyBindingPressEvent<GlobalAction> e)
                 return true;
             }
 
-            if (e.Action == Hotkey)
+            if (e.Action == Hotkey || e.Action == GlobalAction.Select)
             {
                 if (customisationVisible.Value)
                     customisationVisible.Value = false;

```

3stantedja commented:
Opening Settings whilst on the new mods overlay doesn't dim the overlay:
<img width="1595" alt="image" src="https://user-images.githubusercontent.com/3055647/167097128-096a557e-b04e-49d2-a2c5-398ecb738ba2.png">

bdach commented:
As a point of order, I've added all things added thus far that I can fix without needing new design to the OP as a checklist.

frenzibyte commented:
> Opening Settings whilst on the new mods overlay doesn't dim the overlay:
> <img alt="image" width="1595" src="https://user-images.githubusercontent.com/3055647/167097128-096a557e-b04e-49d2-a2c5-398ecb738ba2.png">

I do wonder if that's within scope here, since every other overlay causes the same issue (screenshot below from master):
<img width="1708" alt="CleanShot 2022-05-06 at 21 33 12@2x" src="https://user-images.githubusercontent.com/22781491/167197668-853a6d29-c1f2-4e53-8bef-4aac4fe17b00.png">

3stantedja commented:
That's a fair point; I didn't test with other overlays from master. Might be better addressed with a separate issue?

bdach committed:
Merge branch 'mod-overlay/back-button' into mod-overlay/integration

bdach committed:
Dismiss mod select if select binding is pressed

bdach committed:
Fix incorrect assert in multiplayer song select test scene

bdach committed:
Merge branch 'mod-overlay/back-button' into mod-overlay/integration

bdach committed:
Merge branch 'master' into mod-overlay/integration

bdach commented:
I've merged master to hide commits from dependency PRs, but there are [test failures](https://github.com/ppy/osu/runs/6334112574?check_suite_focus=true) I'm looking into. Something from the recent changes to how the mod column contents are loaded, no doubt.

bdach committed:
Fix invalid `IsLoaded` check

Would fail when `availableMods` would be changed from a different thread
while the columns aren't loaded.

bdach commented:
Pushed a commit that should hopefully resolve test failures (it does locally at least).

Going to tentatively unblock this pull. Not all concerns raised were resolved yet, but I feel that the remaining ones are minor and can be addressed in follow-up PRs.

peppy commented on osu.Game/Overlays/Mods/ModSelectScreen.cs:
```diff
@@ -50,6 +49,15 @@ public abstract class ModSelectScreen : ShearedOverlayContainer
+        /// This is handled locally here rather than via <see cref="FooterButtonMods"/>, because this overlay is being registered at the game level
+        /// and therefore takes away keyboard focus from the screen stack.
+        /// </remarks>
+        internal GlobalAction? Hotkey { get; set; }
```
Is there a case where this isn't set?

bdach replied:
In multiplayer, `RoomSubScreen` to be exact. Would be weird to have F1 work as a toggle binding there I think?

frenzibyte replied:
On testing, this can actually get weird with freemod select overlay, since on master, pressing F1 while on the freemod overlay would close it and open the regular one. While here, pressing F1 will instead just switch to the previous track:

https://user-images.githubusercontent.com/22781491/167262299-814d509d-ad5c-480d-ae10-3e975e091f0c.mp4

Not sure if hardcoding will help or make things worse, though.

peppy replied:
I actually think I'd prefer to have it work there. I've heard some people complain they can't find the local mods button, and the first thing I'd try as a user in that scenario is hitting F1.

It doesn't look to conflict, so I think i'll hardcode this?

peppy replied:
I guess it won't actually work though, funnily enough (only works in the outwards direction). Kinda weird.

bdach replied:
The footer buttons are handling open. The mod select itself is only locally handling close, because it is a focused overlay that takes keyboard focus away from the screen behind it.

peppy replied:
I'm going to hard code this anyway. Will track the showing part at https://github.com/ppy/osu/issues/18140.

peppy committed:
Hardcode hide key handling in `ModSelectScreen`

bdach commented on osu.Game/Overlays/Mods/ModSelectScreen.cs:
```diff
@@ -384,7 +384,7 @@ public override bool OnPressed(KeyBindingPressEvent<GlobalAction> e)
+            if (e.Action == Hotkey || e.Action == GlobalAction.Select)
             {
                 if (customisationVisible.Value)
                     customisationVisible.Value = false;
```
Not sure that this is something that ever realistically needs addressing...

peppy replied:
Might make sense to require two presses to dismiss. Will see how that feels.

bdach replied:
It does already take two presses to dismiss, as far as I can tell (and I just checked). I believe this review conversation is only talking about the fact that if you press twice quickly and slow down the transition it looks ever so slightly off.

bdach replied:
Actually I take it back, I missed that this was done with the "select" press, i.e. enter, rather than escape.

peppy replied:
It doesn't if you use "select" or "toggle", only back. But I think it feels better having these behave differently.. having to press twice feels a bit slow.

frenzibyte replied:
My concern is more about the way it animates when closing while the mod settings area is open:

https://user-images.githubusercontent.com/22781491/167263722-fce85695-a271-4d39-bc63-e8b2ae6f8f28.mp4

(it feels like the mod settings area should be hard-coupled with the footer, to transition smoothly)

But as I mentioned, this is only a small note for follow-up, not something which needs to be resolved in this PR.

peppy committed:
Add xmldoc for `IsValidMod`

peppy committed:
Reorder methods / properties in `ModSelectScreen` for legibility

peppy committed:
Add various xmldoc for nested classes

peppy commented:
I've done some reordering and added some extra xmldoc as I worked through reviewing this. Going to get this in as-is. Code-wise things are pretty good, but there are various UX issues that have been pointed out. Addressing these individually in smaller PRs is the best way forward.

bdach commented:
Latest changes look OK to me.

TRRESTGHGYTCFTGBCXSZD commented:
It does disappear and does not appear, That's a critical bug that needs to be fixed.
![osu_2022-05-11_17-31-51](https://user-images.githubusercontent.com/40109698/167805742-1252fb7d-8a7d-49a0-a7fd-63423ce5792b.jpg)
![osu_2022-05-11_17-32-05](https://user-images.githubusercontent.com/40109698/167805755-145a6585-e3fc-4700-ac7a-63b06772b916.jpg)
![osu_2022-05-11_17-32-59](https://user-images.githubusercontent.com/40109698/167805757-cfb3644e-5df9-4907-bfce-998aeb3c6861.jpg)

frenzibyte commented:
Please open an issue.
